polkadot/consensus/src/evaluation.rs

```diff
@@ -0,0 +1,135 @@
+		}
+		ParachainOutOfOrder {
+			description("Proposal included parachains out of order."),
+			display("Proposal included parachains out of order."),
+		}
+		UnknownParachain(id: ParaId) {
+			description("Proposal included unregistered parachain."),
+			display("Proposal included unregistered parachain {:?}", id),
+		}
+		WrongParentHash(expected: HeaderHash, got: HeaderHash) {
+			description("Proposal had wrong parent hash."),
+			display("Proposal had wrong parent hash. Expected {:?}, got {:?}", expected, got),
+		}
+		WrongNumber(expected: BlockNumber, got: BlockNumber) {
+			description("Proposal had wrong number."),
+			display("Proposal had wrong number. Expected {:?}, got {:?}", expected, got),
+		}
+		ProposalTooLarge(size: usize) {
+			description("Proposal exceeded the maximum size."),
+			display(
+				"Proposal exceeded the maximum size of {} by {} bytes.",
+				MAX_TRANSACTIONS_SIZE, MAX_TRANSACTIONS_SIZE.saturating_sub(*size)
+			),
+		}
+	}
+}
+
+/// Attempt to evaluate a substrate block as a polkadot block, returning error
+/// upon any initial validity checks failing.
+pub fn evaluate_initial(
+	proposal: &SubstrateBlock,
+	now: Timestamp,
+	parent_hash: &HeaderHash,
+	parent_number: BlockNumber,
+	active_parachains: &[ParaId],
+) -> Result<PolkadotBlock> {
+	const MAX_TIMESTAMP_DRIFT: Timestamp = 60;
+
+	let encoded = Slicable::encode(proposal);
+	let proposal = PolkadotGenericBlock::decode(&mut &encoded[..])
+		.and_then(|b| PolkadotBlock::from(b).ok())
+		.ok_or_else(|| ErrorKind::ProposalNotForPolkadot)?;
+
+	let transactions_size = proposal.extrinsics.iter().fold(0, |a, tx| {
+		a + Slicable::encode(tx).len()
+	});
+
+	if transactions_size > MAX_TRANSACTIONS_SIZE {
+		bail!(ErrorKind::ProposalTooLarge(transactions_size))
+	}
+
+	if proposal.header.parent_hash != *parent_hash {
+		bail!(ErrorKind::WrongParentHash(*parent_hash, proposal.header.parent_hash));
+	}
+
+	if proposal.header.number != parent_number + 1 {
+		bail!(ErrorKind::WrongNumber(parent_number + 1, proposal.header.number));
+	}
+
+	let block_timestamp = proposal.timestamp();
+
+	// lenient maximum -- small drifts will just be delayed using a timer.
+	if block_timestamp > now + MAX_TIMESTAMP_DRIFT {
+		bail!(ErrorKind::TimestampInFuture)
+	}
+
+	{
+		let n_parachains = active_parachains.len();
+		if proposal.parachain_heads().len() > n_parachains {
+			bail!(ErrorKind::TooManyCandidates(n_parachains, proposal.parachain_heads().len()));
+		}
+
+		let mut last_id = None;
+		let mut iter = active_parachains.iter();
+		for head in proposal.parachain_heads() {
+			// proposed heads must be ascending order by parachain ID without duplicate.
+			if last_id.as_ref().map_or(false, |x| x >= &head.parachain_index) {
+				bail!(ErrorKind::ParachainOutOfOrder);
+			}
+
+			if iter.position(|x| x == &head.parachain_index).is_none() {
+				// must be unknown since active parachains are always sorted.
```

Comment:
comment is a bit confusing. The code does not care if `active_parachains` is sorted or not.